### PR TITLE
Improve: punning default arg with type constraint

### LINKED ISSUE #295: Formatting function arguments with default values
I try to format this code:
```
let temp_function ?(value: string = "OMG") = print_endline value
```
Ocamlformat changes it to:
```
let temp_function ?value:((value: string) = "OMG") = print_endline value
```
Why is the argument `value` formatted into a more complicated syntax? I think it would be better to keep it as it is.

### PR DESCRIPTION
**Pull request by jberdine**
close #295

### CODE REVIEW

File: src/Fmt_ast.ml [hunk 1]
```diff
@@ -1034,6 +1034,19 @@ and fmt_fun_args c ?(pro= fmt "") args =
         Cmts.fmt c.cmts loc @@ cbox 0 (fmt "?" $ str l)
     | Val (Optional l, xpat, None) ->
         cbox 0 (fmt "?" $ str l $ fmt ":" $ fmt_pattern c xpat)
+    | Val
+        ( Optional l
+        , { ast=
+              { ppat_desc=
+                  Ppat_constraint ({ppat_desc= Ppat_var {txt; loc}}, typ)
+              ; ppat_attributes= [] } as ast }
```
Comment (hhugo) [inline src/Fmt_ast.ml]: what about comments for ppat_loc ?

Comment (jberdine) [reply src/Fmt_ast.ml]: Those and several others were missing

File: src/Fmt_ast.ml [hunk 2]
```diff
@@ -1022,24 +1034,45 @@ and fmt_fun_args c ?(pro= fmt "") args =
                         ( { ppat_desc= Ppat_var {txt; loc}
                           ; ppat_attributes= [] }
```
Comment (hhugo) [inline src/Fmt_ast.ml]: loc missing here (`ppat_loc`)

File: src/Fmt_ast.ml [hunk 3]
```diff
@@ -1022,24 +1034,45 @@ and fmt_fun_args c ?(pro= fmt "") args =
...
+        Cmts.fmt c.cmts ppat_loc @@ Cmts.fmt c.cmts loc
+        @@ cbox 0 (fmt "?" $ str l)
     | Val (Optional l, xpat, None) ->
         cbox 0 (fmt "?" $ str l $ fmt ":" $ fmt_pattern c xpat)
     | Val
         ( Optional l
-        , {ast= {ppat_desc= Ppat_var {txt; loc}; ppat_attributes= []}}
+        , { ast=
+              { ppat_desc=
+                  Ppat_constraint ({ppat_desc= Ppat_var {txt; loc}}, typ)
```
Comment (hhugo) [inline src/Fmt_ast.ml]: same here

File: src/Fmt_ast.ml (hunk 2 above)
Comment (jberdine) [reply src/Fmt_ast.ml]: Time to enable warning 9 I guess.

File: src/Fmt_ast.ml [hunk 4]
```diff
@@ -998,48 +998,93 @@ and fmt_pattern c ?pro ?parens ({ctx= ctx0; ast= pat} as xpat) =
...
+                    ( {ppat_desc= Ppat_var {txt; loc= loc1}; ppat_loc= loc2}
+                    , typ )
+              ; ppat_loc= loc3
+              ; ppat_attributes= [] } as ast }
         , Some xexp )
       when String.equal l txt ->
-        Cmts.fmt c.cmts loc
+        Cmts.fmt c.cmts loc3 @@ Cmts.fmt c.cmts loc2 @@ Cmts.fmt c.cmts loc1
+        @@ cbox 0
+             ( fmt "?(" $ str l $ fmt " :@ "
```
Comment (hhugo) [inline src/Fmt_ast.ml]: I think comments are moving too much.
```
let f ?((* a *) x (* b *) = (* c *) f x y) = e
```
 -> 
```
let f (* a *) ?(x= (* c *) f x y) = (* b *) e
```

**Pull request merged**